Update dbt-trino's connection and configuration docs (#3055)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-configs/trino-configs.md

---

> title: "Starburst (Trino) configurations"

Reviewer: in my pr #3098 , i used `Connect Starburst/Trino` with a `/` instead of `()`. 

i don't have strong feels for either way. whatever we want to go with, we should be consistent.

Author: ```suggestion
title: "Starburst/Trino configurations"
```

---
title: "Starburst (Trino) configurations"
id: "trino-configs"
---

## Model-level configuration

### Session properties

> In some specific cases, there may be needed tuning through the Trino session properties only for a specific dbt model. In such cases, using the [dbt hooks](https://docs.getdbt.com/reference/resource-configs/pre-hook-post-hook) may come to the rescue:

Reviewer 2: ```suggestion
In some specific cases, there may be need for tuning during the Trino session properties only for a specific dbt model. In such cases, using the [dbt hooks](https://docs.getdbt.com/reference/resource-configs/pre-hook-post-hook) may come to the rescue:
```

```sql
{{
  config(

>     pre_hook="set session query_max_run_time='10m'"

Reviewer 2: Since it's in the session, should this be a SQL header use case? https://docs.getdbt.com/reference/resource-configs/sql_header

Author: huh i've always wondered what SQL headers are for... though i'm still confused and I don't think Chat GPT is right here...?
<img width="968" alt="image" src="https://user-images.githubusercontent.com/8158673/228617415-05b5e7ed-0410-4779-9721-849e13370cf8.png">

Reviewer 2: That looks wrong... hm I always thought we wanted things to happen in the session to be SQL headers but I might be wrong.

  )
}}
```

### Use table properties to configure connector specifics

Trino connectors use table properties to configure connector specifics.

> Check the Trino connector documentation for more information.

Reviewer 2: As someone who doesn't know trino - would I be correct that the user needs to check out the specific Trino connector for context (and is this in the context of writing using the connector?).

Reviewer 3: That's correct, connectors differ from each other, including writing support.

Author: ```suggestion
For more information, refer to either the [Trino Connectors](https://trino.io/docs/current/connector.html) or [Starburst Catalog](https://docs.starburst.io/starburst-galaxy/catalogs/) so learn what is and isn't supported for each underlying data platform
```

```sql
{{
  config(
    materialized='table',

>     properties= {

Reviewer 2: this might be a dumb question but are there are a list of potential properties or is that specific to the connector? --> related to the past question

Reviewer 3: It's available in Starburst or Trino docs.

Author: is it though? each Starburst catalog doc seems to be a tutorial on how to connect not what properties are available to be set and their defatuls. Correct me if I'm wrong though @hovasco!

ok this is stranger to me these properties seem to not have to do with underlying catalog, but rather the table format that's being used, as in Delta vs Iceberg vs HUDI?
https://docs.starburst.io/starburst-galaxy/sql/great-lakes.html

Reviewer 3: In the case of Great Lakes it's correct, I will bring docs ambiguities internally.

      "format": "'PARQUET'",
      "partitioning": "ARRAY['bucket(id, 2)']",
    }
  )
}}
```

## Materialization-specific configs

### Seeds

> Seeds are CSV files in your dbt project (typically in your data directory), that dbt can load into your data warehouse using the dbt seed command.

Reviewer 2: Could we link to the seeds docs?

Author: ```suggestion
reference: [dbt docs: Seeds](https://docs.getdbt.com/docs/build/seeds)
```

> For dbt-trino batch_size is defined in macro `trino__get_batch_size()` and default value is `1000`.

Reviewer 2: What is the unit of measure for the default value? Like 1000 rows?

Author: yeah this is "cells" which is why this parameter needs to exist

Reviewer 2: Ah awesome, can we add that to the doc?

In order to override default value define within your project a macro like the following:

```jinja2
{% macro default__get_batch_size() %}
  {{ return(10000) }}
{% endmacro %}
```

### Table

> `dbt-trino` supports two modes in `table` materialization `rename` and `drop` configured using `on_table_exists`.

Reviewer 2: ```suggestion
`dbt-trino` supports two modes in `table` materialization: `rename` and `drop` configured using `on_table_exists`.
```

- `rename` - creates intermediate table, then renames the target to backup one and renames intermediate to target one.
- `drop` - drops and recreates a table. It overcomes table rename limitation in AWS Glue.

> By default `table` materialization uses `on_table_exists = 'rename'`, see an examples below how to change it.

Reviewer 2: Are there recommendations on when to use either?

Reviewer 3: Use a default `rename` wherever possible.

Author: ```suggestion
The recommended `table` materialization uses `on_table_exists = 'rename'` and it is also the default. See the below example for how to change it.
```

In model add:

```jinja2
{{
  config(
    materialized = 'table',
    on_table_exists = 'drop`
  )
}}
```

or in `dbt_project.yaml`:

```yaml
models:
  path:
    materialized: table
    +on_table_exists: drop
```

> Using `table` materialization and `on_table_exists = 'rename'` with AWS Glue may result in below error:

Reviewer 2: Should we have a troubleshooting section below and put this there? Since it's specific to AWS Glue?

```sh
TrinoUserError(type=USER_ERROR, name=NOT_SUPPORTED, message="Table rename is not yet supported by Glue service")
```

### View

> Adapter supports two security modes in `view` materialization `DEFINER` and `INVOKER` configured using `view_security`.

Reviewer 2: ```suggestion
Adapter supports two security modes in `view` materialization `definer` and `invoker` configured using `view_security`.
```

Lowercase this since it was lowercase above for tables

Set the `+views_enabled` to `false` if your connector doesn't support views.

#### `append` (default)

The default incremental strategy is `append`. `append` only adds the new records based on the condition specified in the `is_incremental()` conditional block.

```jinja2
{{
    config(
      materialized = 'incremental')
}}
select * from {{ ref('events') }}
{% if is_incremental() %}
  where event_ts > (select max(event_ts) from {{ this }})
{% endif %}
```

#### `delete+insert`

Through the `delete+insert` incremental strategy, you can instruct dbt to use a two-step incremental approach. It will first delete the records detected through the configured `is_incremental()` block and re-insert them.

```jinja2
{{
    config(
      materialized = 'incremental',
      unique_key='user_id',
      incremental_strategy='delete+insert',
      )
}}
select * from {{ ref('users') }}
{% if is_incremental() %}
  where updated_ts > (select max(updated_ts) from {{ this }})
{% endif %}
```

#### `merge`

> Through the `merge` incremental strategy, dbt-trino constructs a [`MERGE` statement](https://trino.io/docs/current/sql/merge.html) which `INSERT`s new and `UPDATE`s existing records based on the unique key (specified by `unique_key`).  

Reviewer 2: ```suggestion
Through the `merge` incremental strategy, dbt-trino constructs a [`MERGE` statement](https://trino.io/docs/current/sql/merge.html) which `insert`s new and `update`s existing records based on the unique key (specified by `unique_key`).  
```

> If `unique_key` is not unique `delete+insert` strategy can be used.

Reviewer 2: ```suggestion
If `unique_key` is not unique, `delete+insert` strategy can be used.
```

NOTE that this functionality works on incremental models that use partitioning:

```jinja2
{{
    config(
        materialized = 'incremental',
        properties={
          "format": "'PARQUET'",
          "partitioned_by": "ARRAY['day']",
        }
    )
}}
```

### Materialized view

The adapter also supports [materialized views](https://trino.io/docs/current/sql/create-materialized-view.html).
At every subsequent `dbt run` command, the materialized view is [refreshed](https://trino.io/docs/current/sql/refresh-materialized-view.html).

You can also define custom properties for the materialized view through the `properties` config.

This materialization supports the [full_refresh](https://docs.getdbt.com/reference/resource-configs/full_refresh) config and flag.
Whenever you want to rebuild your materialized view, e.g. when changing underlying SQL query, run `dbt run --full-refresh`.

In model add:

```jinja2
{{
  config(
    materialized = 'materialized_view',
    properties = {
      'format': "'PARQUET'"
    },
  )
}}
```

or in `dbt_project.yaml`:

```yaml
models:
  path:
    materialized: materialized_view
    properties:
      format: "'PARQUET'"
```

### Snapshots

> Commonly, analysts need to "look back in time" at some previous state of data in their mutable tables. While some source data systems are built in a way that makes accessing historical data possible, this is often not the case. dbt provides a mechanism, snapshots, which records changes to a mutable table over time.

Reviewer 2: I think we could just link to the snapshot docs. I think the note about the timestamps should be kept but we don't need to explain snapshots in this way when it's already on our docs

Snapshots implement type-2 Slowly Changing Dimensions over mutable source tables. These Slowly Changing Dimensions (or SCDs) identify how a row in a table changes over time. Imagine you have an orders table where the status field can be overwritten as the order is processed. [See also the dbt docs about snapshots](https://docs.getdbt.com/docs/building-a-dbt-project/snapshots).

An example is given below.

```jinja2
{% snapshot orders_snapshot %}
{{
    config(
        target_database='analytics',
        target_schema='snapshots',
        unique_key='id',
        strategy='timestamp',
        updated_at='updated_at',
    )
}}
select * from {{ source('jaffle_shop', 'orders') }}
{% endsnapshot %}

> ```

Author: ```suggestion
reference: [dbt docs: Snapshots](https://docs.getdbt.com/docs/build/snapshots)
```

Note that the Snapshot feature depends on the `current_timestamp` macro. In some connectors the standard precision (`TIMESTAMP(3) WITH TIME ZONE`) is not supported by the connector eg. Iceberg.

If necessary, you can override the standard precision by providing your own version of the `trino__current_timestamp()` macro as in following example:

```jinja2
{% macro trino__current_timestamp() %}
    current_timestamp(6)
{% endmacro %}
```

## Other Trino/Starburst Relevant Configs

### Prepared statements

> The `dbt seed` feature uses [Trino's prepared statements](https://trino.io/docs/current/sql/prepare.html).

Reviewer 2: Should this not be with our seeds section?
